### ISSUE TITLE
Is this possible?

### ISSUE THREAD

**Issue by frederikhors** [stefandanaita/grpc-build#38]
@stefandanaita thanks for your work.

Do you think this can help me with https://github.com/hyperium/tonic/issues/1386?

**Comment by stefandanaita** [comment]
I don't think it helps given your current setup, but it would probably be a good idea to separate your contracts from the workspace members, in which case you would be able to use `grpc-build` to improve your ergonomics.

**Comment by frederikhors** [comment]
I don't understand what you mean.

Do yo mean to have `proto` dir with:

- player.proto
- team.proto
- and_so_on.proto

instead of:

- player/player.proto
- team/team.proto

?

**Comment by stefandanaita** [comment]
Hmm, not exactly. I've looked at your dir structure in the tonic issue you linked above and it looks like you have separate cargo workspaces for all your protos, including the `common.proto`.

I was suggesting this structure
```
- Cargo.toml (workspace)
- src
  - protos
    - Cargo.toml
    - build.rs
    - src
      - common.proto
      - players
        - players.proto
      - teams
        - teams.proto
  - other application code workspaces, if needed at all
```

If you do it this way, importing the common protos into your teams or players becomes a matter of `import "<replace with package name in common.proto>/common.proto";`

Unless you have a specific reason for which you need the protos to be separate into different cargo workspaces, the one workspace approach makes it much simpler and easier to iterate on.

**Comment by isosphere** [comment]
> Unless you have a specific reason for which you need the protos to be separate into different cargo workspaces, the one workspace approach makes it much simpler and easier to iterate on.

Use case: I am modularizing by extracting services I reuse. For example, I have a crate that has some common protoc types and another for authentication. 

With this architecture[^1] separate crates have proto dependencies between them. 

I am pushing forward by exporting the common proto files as strings and writing these to files in my build scripts to `OUT_DIR`, which I am adding to the include path for `tonic_prost_build`.

[extern_path](https://docs.rs/tonic-prost-build/latest/tonic_prost_build/struct.Builder.html#method.extern_path) looks interesting.

I might have another problem when I have two crates independently using the proto files from the common crate. 

[^1]: an alternative arch would make a fat template in one workspace instead, however, this becomes problematic when you fix a problem/improve something in one of them. With the proposed architecture this is handled with semver.
